src/test/java/org/opentripplanner/middleware/controllers/api/ManageLegTraversalTest.java

```diff
@@ -0,0 +1,306 @@
+            if (outcome.containsKey(tripStatus.name())) {
+                Integer hits = outcome.get(tripStatus.name());
+                hits++;
+                outcome.put(tripStatus.name(), hits);
+            } else {
+                outcome.put(tripStatus.name(), 1);
+            }
+        }
+        for (TripStatus tripStatus : TripStatus.values()) {
+            System.out.println(tripStatus.name() + ": " + outcome.get(tripStatus.name()));
+        }
+        System.out.println("Total: " + traces.length);
+    }
+
+    // WIP.
+    @Test
+    void speedCheck() throws IOException {
+        String traceTimes = CommonTestUtils.getTestResourceAsString("controllers/api/bus-stop-justice-center-trace-times.csv");
+        String[] traces = traceTimes.split("\n");
+        for (int i=0; i<traces.length-1; i++) {
+            String[] traceValuesPoint1 = traces[i].split(",");
+            String[] traceValuesPoint2 = traces[i+1].split(",");
+            ZonedDateTime timePoint1 = ZonedDateTime.parse(traceValuesPoint1[0]);
+            Coordinates coordinates1 = new Coordinates(Double.parseDouble(traceValuesPoint1[1]),Double.parseDouble(traceValuesPoint1[2]));
+            ZonedDateTime timePoint2 = ZonedDateTime.parse(traceValuesPoint2[0]);
+            Coordinates coordinates2 = new Coordinates(Double.parseDouble(traceValuesPoint2[1]),Double.parseDouble(traceValuesPoint2[2]));
+            double dist = getDistance(coordinates1, coordinates2);
+            long seconds = Duration.between(timePoint1,timePoint2).getSeconds();
+            if (seconds == 0) {
+                continue;
+            }
+            double speed = dist / seconds;
+            System.out.println(speed + " : " + dist + " : " + seconds);
+        }
+    }
+
+
+    @Test
+    void canTrackTripWithoutDeviating() {
+        for (int legId = 0; legId < monitoredTrip.itinerary.legs.size(); legId++) {
+            List<ManageLegTraversal.Segment> segments = createSegmentsForLeg(legId);
+            TrackedJourney trackedJourney = new TrackedJourney();
+            trackedJourney.tripId = monitoredTrip.id;
+            ZonedDateTime startOfTrip = ZonedDateTime.ofInstant(
+                monitoredTrip.itinerary.legs.get(legId).startTime.toInstant(),
+                DateTimeUtils.getOtpZoneId()
+            );
+
+            ZonedDateTime currentTime = startOfTrip;
+            double cumulativeTravelTime = 0;
+            for (ManageLegTraversal.Segment segment : segments) {
+                trackedJourney.locations = List.of(
+                    new TrackingLocation(
+                        segment.coordinates.lat,
+                        segment.coordinates.lon,
+                        new Date((currentTime.toInstant().toEpochMilli()))
+                    )
+                );
+                assertEquals(
+                    ManageTripTracking.getTripStatus(trackedJourney, monitoredTrip).name(),
+                    TripStatus.ON_TRACK.name()
+                );
```

Comment:
This piece of code should be the subject of a parameterized test for a given leg and leg geometry. Basically, for test params you want to pass a timestamp, a position, and an expected status.